File: net/JetBrains.Serialization/src/Dataholders/AttributeInfos/CounterSignatureAttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;
using JetBrains.Serialization;

[JsonObject(MemberSerialization.Fields)]
public class CounterSignatureAttributeInfo : AttributeInfo
{
  public override TextualInfo Identifier { get; }
  public List<CounterSignatureInfo> Content { get; }

  public CounterSignatureAttributeInfo(Attribute attribute)
  {
    Identifier = TextualInfo.GetInstance(attribute.AttrType);
    Content = attribute.AttrValues.ToArray().Select(item => CounterSignatureInfo.GetInstance(item as DerSequence))
      .ToList();
  }

  public override Asn1Encodable GetPrimitiveContent() => Content.ToPrimitiveDerSet();
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/attributeInfos/CounterSignatureAttributeInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toPrimitiveDLSet
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.cms.Attribute

@Serializable
data class CounterSignatureAttributeInfo(
  override val identifier: TextualInfo,
  val content: List<CounterSignatureInfo>
) : AttributeInfo() {

  companion object {
    fun getInstance(attribute: Attribute): CounterSignatureAttributeInfo {
      return CounterSignatureAttributeInfo(
        TextualInfo.getInstance(attribute.attrType),
        attribute.attributeValues.map { CounterSignatureInfo.getInstance(it as DLSequence) }
      )
    }
  }

  override fun getPrimitiveContent() = content.toPrimitiveDLSet()
}